packages/backend/src/model/AssetBalances.ts

```diff
@@ -0,0 +1,43 @@
+interface AssetBalance {
+  assetId: string
+  balance: bigint
+}
+
+export class AssetBalances extends Array<AssetBalance> {
```

Comment:
I am not sure if this is the way I'd go about solving this problem.

Why not have toRow and toRecord stringify and parse the values and not have a special type that extends Array? Extending Array is problematic, because all methods return regular arrays not your custom type.

Reply:
> Extending Array is problematic, because all methods return regular arrays not your custom type.

I remember that was the case in old JavaScript, but I don't think it's the case anymore.

**Node 14**

![image](https://user-images.githubusercontent.com/15332326/148390543-c4867b77-4713-4250-a8be-42536c941614.png)

**Browser (Brave)**

![image](https://user-images.githubusercontent.com/15332326/148390650-5a921eb6-cd8d-47d6-b674-927229486145.png)

Reply:
The root cause of the problem is an inconstistency on Knex and `node-postgres` side.

We need to pass a stringified JSON to the database — https://knexjs.org/#Schema-jsonb.
But, when we receive it from the database, we get the JSON object alraedy _partially_ parsed with `JSON.parse`.

After receiving it, I'm turning all `balance` properties into `bigint` — I'm not quite sure it's needed, but I wanted to preserve the types you wrote in `OnChainData.ts`.

Reply:
I can inline this logic in `toRow` and `toRecord` if you'd prefer.

Another question I have is: Would we ever want to filter or query position updates based on an asset balance? Should this maybe become a table in the database instead?

Reply:
There is also the issue of TypeScript handling of the type:
![image](https://user-images.githubusercontent.com/17070569/148393815-c2bac11c-9aa7-47f8-9ab0-6c82286646f1.png)